Implement more useful workflow tagging

jmchilton opened the pull request:
Extends the workflow index search and pagination functionality added in #13725 to enhance tag functionality. 

- By default now free search terms are also searched against workflow tags.
- Github-style "search tags" as used with the tool shed repository search and TRS searching can no be used for "stored workflow tags" and names - to restrict search terms.
- The workflow tags in the grid themselves can now be clicked and the query textbox will reactively update to include a "search tag" to search by the  "workflow tag".

A grid with tagged workflows:

<img width="1189" alt="Screen Shot 2022-04-21 at 3 47 14 PM" src="https://user-images.githubusercontent.com/216771/164541117-4705aaba-796f-416d-837e-a53857e90131.png">

Once a tag is clicked on:

<img width="1197" alt="Screen Shot 2022-04-21 at 3 47 21 PM" src="https://user-images.githubusercontent.com/216771/164541120-516f8490-c080-45bf-b2fc-09da70278ee3.png">

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] Instructions for manual testing are as follows:
  1. Search workflows by tag.

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

jmchilton committed:
Clarify updating workflow tags in API test.

jmchilton committed:
Improved search for workflows...

Use Github-like search tags like we do for tool shed repositories and TRS attributes.

jmchilton committed:
Add filter by tag by clicking on the tag of workflows in UI.

jmchilton committed:
Typing and testing for baseline galaxy.util.search functionality.

jmchilton committed:
Allow quoting in search library for exact matches (e.g. filtering on specific tags/tools).

jmchilton committed:
Selenium test case for workflow search by name filters.

jmchilton committed:
More workflow tagging tests.

jmchilton committed:
More percise workflow searching.

Allow quoting, exact tag matching, etc...

jmchilton committed:
Improved filtering by tag in the GUI.

davelopez reviewed (approved):
Works great!
Just one thing, It seems the filtering is case sensitive, is that intended?

![search-workflows](https://user-images.githubusercontent.com/46503462/165272211-4fae5527-49fb-427e-b7fc-089d38699f19.gif)

Since this one depends on https://github.com/galaxyproject/galaxy/pull/13725 it will require a rebase after applying the fix there.

mvdbeek commented on client/src/components/Workflow/WorkflowList.vue:
```diff
@@ -143,47 +156,44 @@ export default {
             titleRunWorkflow: _l("Run workflow"),
+            workflowItemsModel: [],
+            workflowItems: [],
+            perPage: this.rowsPerPage(20),
```
```suggestion
            perPage: this.rowsPerPage(50),
```
20 is kind of low, 50 is what the invocations, review cleanup,  and dataset search components use. (And it would be really cool if we could make this a sticky user preference at one point in the future).

jmchilton replied:
I don't agree but I've made the switch in #13725. I think the workflow list is more like a Google search result. I'm imaging a future with rich text descriptions and workflow icons and box-y instead of grid display (ala discussed in the context of simplified workflow interface and seen on say https://nf-co.re/pipelines). In that context - Google's 20 feels right. Invocations/jobs/datasets are more like proper rows in a large table to me and the limit of 50 feels right there.

I think the simplified workflow UI placing the workflows above the invocations also influenced me in thinking they should not consume so much space... but I will admit all of those are future concerns.

jmchilton replied:
Rebased into this PR.

jmchilton committed:
Make workflow search case insensitive under Postgres.

Only clean way to get the same behavior between sqlite and Postgres. Probably the preferable thing from the GUI perspective also.

bgruening reviewed (approved):
Thanks @jmchilton! 

I found a few things that could be improved in a follow-up by someone, not strictly you, maybe as paper cuts. I have written it down at https://github.com/galaxyproject/galaxy/issues/13841

I think we should merge it as it is a big improvement over the old one. Thanks again.
